File: .skills/openclaw-skills/skills/0xnyk/xint/lib/mcp_dispatcher.ts
import * as api from "./api";
import * as cache from "./cache";
import { checkBudget, getCostSummary, trackCost } from "./costs";
import { fetchTrends, resolveWoeid } from "./trends";
import { fetchArticle, fetchTweetForArticle } from "./article";
import { extractTweetId } from "./media";
import { actionInfo, actionSuccess, type ActionExecutionResult } from "./action_result";

export type ToolExecutionResult = ActionExecutionResult<unknown>;

export type MCPToolHandler = (args: Record<string, unknown>) => Promise<ToolExecutionResult>;

export type MCPDispatcherDeps = {
  extractTweetId: (input: string) => string;
  callPackageApi: (method: string, path: string, body?: unknown) => Promise<unknown>;
  ensurePackageQueryCitations: (data: unknown, requireCitations: boolean) => void;
};

export function createMcpToolHandlers(deps: MCPDispatcherDeps): Record<string, MCPToolHandler> {
  return {
    async xint_search(args) {
      const query = String(args.query || "");
      const tweets = await api.search(query, {
        pages: Math.ceil((Number(args.limit) || 15) / 20),
        sortOrder: (args.sort === "recent" ? "recency" : "relevancy") as any,
        since: typeof args.since === "string" ? args.since : undefined,
      });

      let results = tweets;
      if (args.no_retweets ?? args.noRetweets) {
        results = results.filter((t: any) => !t.text.startsWith("RT @"));
      }
      if (args.no_replies ?? args.noReplies) {
        results = results.filter((t: any) => t.conversation_id === t.id);
      }
      trackCost("search", "/2/tweets/search/recent", tweets.length);
      const limit = Number(args.limit) || 15;
      const sliced = results.slice(0, limit);
      const result = actionSuccess("Search completed.", sliced);
      result.pagination = { total: results.length, returned: sliced.length, has_more: results.length > limit };
      result.cost = tweets.length * 0.005;
      return result;
    },

    async xint_profile(args) {
      const username = String(args.username || "");
      const count = Number(args.count) || 20;
      const includeReplies = Boolean(args.include_replies ?? args.includeReplies);
      const { user, tweets } = await api.profile(username, { count, includeReplies });
      trackCost("profile", `/2/users/by/username/${username}`, tweets.length + 1);
      const slicedTweets = tweets.slice(0, count);
      const result = actionSuccess("Profile lookup completed.", { user, tweets: slicedTweets });
      result.pagination = { total: tweets.length, returned: slicedTweets.length, has_more: tweets.length > count };
      result.cost = (tweets.length + 1) * 0.005;
      return result;
    },

    async xint_thread(args) {
      const tweetId = deps.extractTweetId(String(args.tweet_id || args.tweetId || ""));
      const pages = Number(args.pages) || 2;
      const tweets = await api.thread(tweetId, { pages });
      trackCost("thread", "/2/tweets/search/recent", tweets.length);
      return actionSuccess("Thread lookup completed.", { tweets });
    },

    async xint_tweet(args) {
      const tweetId = deps.extractTweetId(String(args.tweet_id || args.tweetId || ""));
      const tweet = await api.getTweet(tweetId);
      trackCost("tweet", `/2/tweets/${tweetId}`, tweet ? 1 : 0);
      return actionSuccess("Tweet lookup completed.", tweet);
    },

    async xint_article(args) {
      let articleUrl = String(args.url || "");
      const full = args.full !== false;

      // Check if it's a tweet URL — may contain an inline X Article
      if (extractTweetId(articleUrl)) {
        const ctx = await fetchTweetForArticle(articleUrl);
        if (ctx.inlineArticle) {
          return actionSuccess("Article fetched from X Article data.", ctx.inlineArticle);
        }
        if (ctx.articleUrl) {
          articleUrl = ctx.articleUrl;
        } else {
          throw new Error("No article link found in tweet");
        }
      }

      const article = await fetchArticle(articleUrl, { full });
      return actionSuccess("Article fetch completed.", article);
    },

    async xint_xsearch() {
      return actionInfo("xSearch requires XAI_API_KEY.", { note: "xSearch requires XAI_API_KEY" });
    },

    async xint_collections_list() {
      return actionInfo("Collections requires XAI_API_KEY.", { note: "Collections requires XAI_API_KEY" });
    },

    async xint_collections_search() {
      return actionInfo("Collections requires XAI_API_KEY.", { note: "Collections requires XAI_API_KEY" });
    },

    async xint_analyze() {
      return actionInfo("Analyze requires XAI_API_KEY.", { note: "Analyze requires XAI_API_KEY" });
    },

    async xint_trends(args) {
      const location = typeof args.location === "string" ? args.location : "worldwide";
      const limit = Number(args.limit) || 20;
      const trends = await fetchTrends(resolveWoeid(location));
      return actionSuccess(
        "Trends fetch completed.",
        { ...trends, trends: trends.trends.slice(0, limit) },
        trends.source === "search_fallback",
      );
    },

    async xint_bookmarks() {
      return actionInfo("Bookmarks requires OAuth - use xint bookmarks command.", {
        note: "Bookmarks requires OAuth - use xint bookmarks command",
      });
    },

    async xint_package_create(args) {
      const payload = {
        name: String(args.name || ""),
        topic_query: String(args.topicQuery || args.topic_query || ""),
        sources: Array.isArray(args.sources) ? args.sources : [],
        time_window: (args.timeWindow || args.time_window || {
          from: new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString(),
          to: new Date().toISOString(),
        }) as unknown,
        policy: String(args.policy || "private"),
        analysis_profile: String(args.analysisProfile || args.analysis_profile || "summary"),
      };
      const data = await deps.callPackageApi("POST", "/packages", payload);
      return actionSuccess("Package create request accepted.", data);
    },

    async xint_package_status(args) {
      const packageId = String(args.packageId || args.package_id || "");
      if (!packageId) throw new Error("Missing packageId/package_id");
      const data = await deps.callPackageApi("GET", `/packages/${encodeURIComponent(packageId)}`);
      return actionSuccess("Package status fetched.", data);
    },

    async xint_package_query(args) {
      const requireCitations =
        args.requireCitations !== undefined
          ? Boolean(args.requireCitations)
          : args.require_citations !== undefined
            ? Boolean(args.require_citations)
            : true;
      const payload = {
        query: String(args.query || ""),
        package_ids: Array.isArray(args.packageIds)
          ? args.packageIds
          : Array.isArray(args.package_ids)
            ? args.package_ids
            : [],
        max_claims: Number(args.maxClaims || args.max_claims || 10),
        require_citations: requireCitations,
      };
      if (!payload.query || payload.package_ids.length === 0) {
        throw new Error("Missing query or packageIds/package_ids");
      }
      const data = await deps.callPackageApi("POST", "/query", payload);
      deps.ensurePackageQueryCitations(data, requireCitations);
      return actionSuccess("Package query completed.", data);
    },

    async xint_package_refresh(args) {
      const packageId = String(args.packageId || args.package_id || "");
      if (!packageId) throw new Error("Missing packageId/package_id");
      const payload = { reason: String(args.reason || "manual") };
      const data = await deps.callPackageApi(
        "POST",
        `/packages/${encodeURIComponent(packageId)}/refresh`,
        payload,
      );
      return actionSuccess("Package refresh requested.", data);
    },

    async xint_package_search(args) {
      const query = String(args.query || "");
      if (!query) throw new Error("Missing query");
      const limit = Number(args.limit || 20);
      const data = await deps.callPackageApi(
        "GET",
        `/packages/search?q=${encodeURIComponent(query)}&limit=${encodeURIComponent(String(limit))}`,
      );
      return actionSuccess("Package search completed.", data);
    },

    async xint_package_publish(args) {
      const packageId = String(args.packageId || args.package_id || "");
      const snapshotVersion = Number(args.snapshotVersion || args.snapshot_version || 0);
      if (!packageId || !snapshotVersion) {
        throw new Error("Missing packageId/package_id or snapshotVersion/snapshot_version");
      }
      const data = await deps.callPackageApi(
        "POST",
        `/packages/${encodeURIComponent(packageId)}/publish`,
        { snapshot_version: snapshotVersion },
      );
      return actionSuccess("Package publish requested.", data);
    },

    async xint_watch(args) {
      const query = String(args.query || "");
      const limit = Number(args.limit) || 10;
      const since = typeof args.since === "string" ? args.since : "1h";
      const tweets = await api.search(query, {
        pages: 1,
        sortOrder: "recency",
        since,
      });
      trackCost("watch", "/2/tweets/search/recent", tweets.length);
      const watchSliced = tweets.slice(0, limit);
      const watchResult = actionSuccess("Watch check completed.", watchSliced);
      watchResult.pagination = { total: tweets.length, returned: watchSliced.length, has_more: tweets.length > limit };
      watchResult.cost = tweets.length * 0.005;
      return watchResult;
    },

    async xint_diff(args) {
      return actionInfo("Diff requires OAuth - use xint diff command.", {
        note: "Diff requires OAuth - use xint diff command",
      });
    },

    async xint_report(args) {
      const topic = String(args.topic || "");
      const pages = Number(args.pages) || 2;
      const tweets = await api.search(topic, { pages, sortOrder: "relevancy" });
      const sorted = api.sortBy(tweets, "likes");
      trackCost("report", "/2/tweets/search/recent", tweets.length);
      const reportResult = actionSuccess("Report data gathered.", {
        topic,
        tweet_count: sorted.length,
        top_tweets: sorted.slice(0, 20),
      });
      reportResult.cost = tweets.length * 0.005;
      return reportResult;
    },

    async xint_sentiment(args) {
      return actionInfo("Sentiment requires XAI_API_KEY.", { note: "Sentiment requires XAI_API_KEY" });
    },

    async xint_cache_clear() {
      const removed = cache.clear();
      return actionSuccess("Cache cleared.", { cleared: removed });
    },

    async xint_costs(args) {
      const rawPeriod = typeof args.period === "string" ? args.period : "today";
      const period = ["today", "week", "month", "all"].includes(rawPeriod) ? rawPeriod : "today";
      const summary = getCostSummary(period as "today" | "week" | "month" | "all");
      const budget = checkBudget();
      return actionSuccess("Cost summary generated.", { period, summary, budget });
    },
  };
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/mcp_dispatcher.rs
use crate::action_result::{action_error, action_success, ActionExecutionResult};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum McpToolRoute {
    Search,
    Profile,
    Thread,
    Tweet,
    Trends,
    XSearch,
    CollectionsList,
    Analyze,
    Article,
    CollectionsSearch,
    Bookmarks,
    PackageCreate,
    PackageStatus,
    PackageQuery,
    PackageRefresh,
    PackageSearch,
    PackagePublish,
    CacheClear,
    Watch,
    Diff,
    Report,
    Sentiment,
    Costs,
}

pub fn resolve_tool_route(name: &str) -> ActionExecutionResult<McpToolRoute> {
    match name {
        "xint_search" => action_success("tool route resolved", Some(McpToolRoute::Search)),
        "xint_profile" => action_success("tool route resolved", Some(McpToolRoute::Profile)),
        "xint_thread" => action_success("tool route resolved", Some(McpToolRoute::Thread)),
        "xint_tweet" => action_success("tool route resolved", Some(McpToolRoute::Tweet)),
        "xint_trends" => action_success("tool route resolved", Some(McpToolRoute::Trends)),
        "xint_xsearch" => action_success("tool route resolved", Some(McpToolRoute::XSearch)),
        "xint_collections_list" => {
            action_success("tool route resolved", Some(McpToolRoute::CollectionsList))
        }
        "xint_analyze" => action_success("tool route resolved", Some(McpToolRoute::Analyze)),
        "xint_article" => action_success("tool route resolved", Some(McpToolRoute::Article)),
        "xint_collections_search" => {
            action_success("tool route resolved", Some(McpToolRoute::CollectionsSearch))
        }
        "xint_bookmarks" => action_success("tool route resolved", Some(McpToolRoute::Bookmarks)),
        "xint_package_create" => {
            action_success("tool route resolved", Some(McpToolRoute::PackageCreate))
        }
        "xint_package_status" => {
            action_success("tool route resolved", Some(McpToolRoute::PackageStatus))
        }
        "xint_package_query" => {
            action_success("tool route resolved", Some(McpToolRoute::PackageQuery))
        }
        "xint_package_refresh" => {
            action_success("tool route resolved", Some(McpToolRoute::PackageRefresh))
        }
        "xint_package_search" => {
            action_success("tool route resolved", Some(McpToolRoute::PackageSearch))
        }
        "xint_package_publish" => {
            action_success("tool route resolved", Some(McpToolRoute::PackagePublish))
        }
        "xint_cache_clear" => action_success("tool route resolved", Some(McpToolRoute::CacheClear)),
        "xint_watch" => action_success("tool route resolved", Some(McpToolRoute::Watch)),
        "xint_diff" => action_success("tool route resolved", Some(McpToolRoute::Diff)),
        "xint_report" => action_success("tool route resolved", Some(McpToolRoute::Report)),
        "xint_sentiment" => action_success("tool route resolved", Some(McpToolRoute::Sentiment)),
        "xint_costs" => action_success("tool route resolved", Some(McpToolRoute::Costs)),
        _ => action_error(format!("Unknown tool: {name}")),
    }
}

#[cfg(test)]
mod tests {
    use super::{resolve_tool_route, McpToolRoute};

    #[test]
    fn resolves_known_tools() {
        assert_eq!(
            resolve_tool_route("xint_search").data,
            Some(McpToolRoute::Search)
        );
        assert_eq!(
            resolve_tool_route("xint_package_query").data,
            Some(McpToolRoute::PackageQuery)
        );
    }

    #[test]
    fn rejects_unknown_tools() {
        assert!(resolve_tool_route("xint_unknown").data.is_none());
    }
}
